[JSC] Implement `Array.prototype.toSorted` in C++

sosukesuzuki opened the pull request:
#### 2e3b7325657e70052d111fbf53a3bd4e07a484df
<pre>
[JSC] Implement `Array.prototype.toSorted` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=284445">https://bugs.webkit.org/show_bug.cgi?id=284445</a>

Reviewed by Yusuke Suzuki.

In current implementation, `Array#toSorted` is written in JavaScript.
This patch implements it in C++.

In BaselineJIT, this patch makes it 1.2x faster. Also no performance
regression by this patch.

With DFG/FTL:

                                                      TipOfTree                  Patched

    array-prototype-toSorted-large-array-comparator-double
                                                   11.6039+-0.1263     ^     11.0654+-0.2067        ^ definitely 1.0487x faster

    &lt;geometric&gt;                                    11.6039+-0.1263     ^     11.0654+-0.2067        ^ definitely 1.0487x faster

Without DFG/FTL:

                                                      TipOfTree                  Patched

    array-prototype-toSorted-large-array-comparator-double
                                                   14.3067+-0.1834     ^     11.7961+-0.1853        ^ definitely 1.2128x faster

    &lt;geometric&gt;                                    14.3067+-0.1834     ^     11.7961+-0.1853        ^ definitely 1.2128x faster

* JSTests/microbenchmarks/array-prototype-toSorted-large-array-comparator-double.js: Added.
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toSorted): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::sortImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: <a href="https://commits.webkit.org/287734@main">https://commits.webkit.org/287734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2605b7dbf77b376b60b8d00dfddd30ff6eb55f4f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80644 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59651 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85168 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31626 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68712 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62985 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20783 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83713 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73411 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43290 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50397 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27571 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30085 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73624 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71544 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28095 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86601 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79703 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7870 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5551 "Found 2 new test failures: imported/w3c/web-platform-tests/dom/observable/tentative/observable-filter.any.worker.html media/video-canvas-createPattern.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71283 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8045 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69247 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70524 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17576 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14534 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13485 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102110 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7832 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24854 "Found 3 new JSC stress test failures: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager-jettison, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7671 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11190 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9476 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1965,6 +1976,56 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncToReversed, (JSGlobalObject* globalObject
+    }
+
+    sortImpl(globalObject, result, length, comparatorValue);
+    RETURN_IF_EXCEPTION(scope, { });
```
This part should be,

```
scope.release();
sortImpl(globalObject, result, length, comparatorValue);
return JSValue::encode(result);
```

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1179,18 +1169,39 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncSort, (JSGlobalObject* globalObject, Call
+    sortImpl(globalObject, thisObject, length, comparatorValue);
+    RETURN_IF_EXCEPTION(scope, { });
 
     return JSValue::encode(thisObject);
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

sosukesuzuki committed:
[JSC] Implement `Array.prototype.toSorted` in C++
https://bugs.webkit.org/show_bug.cgi?id=284445

Reviewed by Yusuke Suzuki.

In current implementation, `Array#toSorted` is written in JavaScript.
This patch implements it in C++.

In BaselineJIT, this patch makes it 1.2x faster. Also no performance
regression by this patch.

With DFG/FTL:

                                                      TipOfTree                  Patched

    array-prototype-toSorted-large-array-comparator-double
                                                   11.6039+-0.1263     ^     11.0654+-0.2067        ^ definitely 1.0487x faster

    <geometric>                                    11.6039+-0.1263     ^     11.0654+-0.2067        ^ definitely 1.0487x faster

Without DFG/FTL:

                                                      TipOfTree                  Patched

    array-prototype-toSorted-large-array-comparator-double
                                                   14.3067+-0.1834     ^     11.7961+-0.1853        ^ definitely 1.2128x faster

    <geometric>                                    14.3067+-0.1834     ^     11.7961+-0.1853        ^ definitely 1.2128x faster

* JSTests/microbenchmarks/array-prototype-toSorted-large-array-comparator-double.js: Added.
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toSorted): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::sortImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: https://commits.webkit.org/287734@main
